core/src/main/java/com/statful/client/core/buffer/AggregatedBuffer.java

```diff
@@ -0,0 +1,172 @@
+            }
+        }
+
+        aggregatedBuffer.put(aggregationFrequency.toString(), aggregatedFreqBuffer);
+        buffer.put(aggregation.toString(), aggregatedBuffer);
+
+        return aggregatedFreqBuffer.offer(metric);
+    }
+
+    /**
+     * Reads the buffer contents for a particular aggregation.
+     * @param aggregation The {@link Aggregation} aggregation to inspect the buffer
+     * @param aggregationFrequency The {@link AggregationFrequency} aggregation frequency to inspect the buffer
+     * @return A {@link String} with all the metrics for a particular aggregation
+     */
+    public final String readBuffer(final Aggregation aggregation, final AggregationFrequency aggregationFrequency) {
+        Map<String, ArrayBlockingQueue<String>> aggregatedBuffer = buffer.get(aggregation.toString());
+        ArrayBlockingQueue<String> aggregatedFreqBuffer;
+
+        if (aggregatedBuffer != null) {
+            aggregatedFreqBuffer = aggregatedBuffer.get(aggregationFrequency.toString());
+
+            if (aggregatedFreqBuffer != null) {
+                Collection<String> messages = new ArrayList<String>();
+                aggregatedFreqBuffer.drainTo(messages, flushSize);
+
+                StringBuilder sb = new StringBuilder();
+                for (String metric : messages) {
+                    sb.append(metric).append("\n");
+                }
+
+                return sb.toString();
+            }
+        }
+
+        return "";
+    }
+
+    /**
+     * Return the current aggregations buffers.
+     * @return A {@link Set} set with the current aggregations buffers
+     */
+    public final Set<Aggregation> getAggregations() {
+        Set<String> aggregationNames = buffer.keySet();
+        Set<Aggregation> aggregations = new HashSet<Aggregation>();
+
+        for (String aggregation : aggregationNames) {
+            aggregations.add(Aggregation.valueOf(aggregation));
+        }
+
+        return aggregations;
+    }
+
+    /**
+     * Return the current aggregation frequencies buffers.
+     * @param aggregation The {@link Aggregation} aggregation to inspect the buffer
+     * @return A {@link Set} set with the current aggregations frequencies buffers
+     */
+    public final Set<AggregationFrequency> getAggregationFrequencies(final Aggregation aggregation) {
+        Map<String, ArrayBlockingQueue<String>> aggregationFreqBuffer = buffer.get(aggregation.toString());
+
+        Set<AggregationFrequency> aggregationFrequencies = new HashSet<AggregationFrequency>();
+
+        if (aggregationFreqBuffer != null) {
+            for (String aggregationFreq : aggregationFreqBuffer.keySet()) {
+                aggregationFrequencies.add(AggregationFrequency.valueOf(aggregationFreq));
+            }
+        }
+
+        return aggregationFrequencies;
+    }
+
+    @Override
+    public final boolean isTimeToFlush() {
+        Set<Aggregation> aggregations = getAggregations();
```

Comment:
Sorry for not raising this issue yesterday but we were focusing on another problem and this suggestion didn't came up.

Probably I'm not seeing the whole picture but it seems that exists room for reducing the time complexity of this class and specifically this method. My suggestion is to use a single buffer and to prefix the aggregation and aggregation frequency to the metric data while writing to the buffer, instead of using multiple buffers. You could use those prefixes to know what to do. This way, this method will have linear complexity.

This might be a problem if the client has a thousands of metrics to flush and the method we might have to worry about most is the synchronized flush, which runs on the user's threads.

I think it's fine for now, but I think we should refactor this on later iterations.